src/dune_engine/action_builder.ml

```diff
@@ -372,7 +378,9 @@ struct
         let* (x, deps), deps_x = exec t in
         let+ deps = Build_deps.register_action_deps deps in
         (x, merge_facts deps deps_x)
-      | Fail { fail } -> fail ()
+      | Fail { fail } ->
+        let+ () = Memo.Build.return () in
```

Comment:
What's the reason for introducing this delay?

Reply:
To make sure the exception is raised at "monad execute time" rather than at "monad construction time". The latter could cause an error to be raised too eagerly. It didn't used to be an issue before because the only place where we were executing the monad was inside `execute_rule`. But now we do that in more places because of the calls to `Build_system.eval_build_request`

Reply:
We no longer need this change in the end, so I reverted it